schedserver/task.go

```diff
@@ -44,62 +128,130 @@ func getTask(client *clientv3.Client, ctx context.Context, id *api.TaskID) (*Tas
 		log.Fatalln("Too many matching keys, found:", resp.Count)
 	}
 
-	task := &Task{}
+	task := &Task{version: resp.Kvs[0].Version, key: key, Task: &pb.Task{}}
 	if err := proto.Unmarshal([]byte(resp.Kvs[0].Value), task.Task); err != nil {
 		return nil, err
 	}
 
-	task.revision = resp.Kvs[0].ModRevision
+	// Ensure what we requested and what we got back match up
+	if task.Id.Uuid != id.Uuid {
+		return nil, fmt.Errorf("Requested task %s, received %s", id.Uuid, task.Id.Uuid)
+	}
 
 	return task, nil
 }
 
+// taskKey returns the etcd key for a TaskID
 func taskKey(id *api.TaskID) string {
-	return taskPrefix + id.Uuid
+	return idKey(taskFmt, id)
 }
 
+// statusKey returns the etcd status key of a Task for a given TaskStatus
 func (t *Task) statusKey(status *api.TaskStatus) string {
-	key := statusPrefix
+	// Ensures the Task ID is passed in
+	key := func(format string, args ...interface{}) string {
+		return idKey(format, t.Id, args...)
+	}
 
-	// See README/#ETCD Key Schema
 	switch status.Status.(type) {
 	case *api.TaskStatus_Queued_:
-		key += "queued"
+		return key(queuedFmt)
 	case *api.TaskStatus_Running_:
-		key += "running" + keySep + status.GetRunning().NodeId.Uuid
+		return key(runningFmt, status.GetRunning().NodeId.Uuid)
 	case *api.TaskStatus_Complete_:
-		key += "complete" + keySep + strconv.FormatInt(status.GetComplete().Epoch, 10)
+		return key(completeFmt, status.GetComplete().Epoch)
 	case *api.TaskStatus_Canceled_:
-		key += "canceled" + keySep + strconv.FormatInt(status.GetCanceled().Epoch, 10)
+		return key(canceledFmt, status.GetCanceled().Epoch)
+	default:
+		// TODO - Is this wise?
```

Comment:
That was my though aswell - at least it'll consistently fail if this happens.

Reply:
Crashing is fine but it's should be the mindset for the whole team.
I agree that crashing to expose a bug in the software is best since otherwise the bug might spent years buried in the software.
And I think, if there's a tooling in place to sync the new types with everything else (i.e. either automatically providing the implementation of such functions or warning about unused/unknown/errorneous cases), it is safe to panic.